backend/status/status_source.go

```diff
@@ -0,0 +1,213 @@
+package status
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"syscall"
+
+	grpccodes "google.golang.org/grpc/codes"
+	grpcstatus "google.golang.org/grpc/status"
+)
+
+// Source type defines the status source.
+type Source string
+
+const (
+	// SourcePlugin status originates from plugin.
+	SourcePlugin Source = "plugin"
+
+	// SourceDownstream status originates from downstream service.
+	SourceDownstream Source = "downstream"
+
+	// DefaultSource is the default [Source] that should be used when it is not explicitly set.
+	DefaultSource Source = SourcePlugin
+)
+
+// IsValid return true if es is [SourceDownstream] or [SourcePlugin].
+func (s Source) IsValid() bool {
+	return s == SourceDownstream || s == SourcePlugin
+}
+
+// String returns the string representation of s. If s is not valid, [DefaultSource] is returned.
+func (s Source) String() string {
+	if !s.IsValid() {
+		return string(DefaultSource)
+	}
+
+	return string(s)
+}
+
+// SourceFromHTTPError returns a [Source] based on provided error.
+func SourceFromHTTPError(err error) Source {
+	if IsDownstreamHTTPError(err) {
+		return SourceDownstream
+	}
+	return SourcePlugin
+}
+
+// ErrorSourceFromStatus returns a [Source] based on provided HTTP status code.
+func SourceFromHTTPStatus(statusCode int) Source {
+	switch statusCode {
+	case http.StatusMethodNotAllowed,
+		http.StatusNotAcceptable,
+		http.StatusPreconditionFailed,
+		http.StatusRequestEntityTooLarge,
+		http.StatusRequestHeaderFieldsTooLarge,
+		http.StatusRequestURITooLong,
+		http.StatusExpectationFailed,
+		http.StatusUpgradeRequired,
+		http.StatusRequestedRangeNotSatisfiable,
+		http.StatusNotImplemented:
+		return SourcePlugin
+	}
+
+	return SourceDownstream
+}
+
+type errorWithSourceImpl struct {
+	source Source
+	err    error
+}
+
+// DownstreamError creates a new error with status [SourceDownstream].
+func DownstreamError(err error) error {
```

Comment:
Do we want to have a similar logic here from `experimental/errorsource` and let users decide if they want to override the error source in case it exists? https://github.com/grafana/grafana-plugin-sdk-go/blob/main/experimental/errorsource/error_source.go#L52

```
// SourceError returns an error with the source
// If source is already defined, it will return it, or you can override
func SourceError(source backend.ErrorSource, err error, override bool) Error {
	var sourceError Error
	if errors.As(err, &sourceError) && !override {
		return sourceError // already has a source
	}
	return Error{
		source: source,
		err:    err,
	}
}
```

Reply:
👍